// angular-bootstrap-md/esm2015/lib/buttons/buttons.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Renderer2, ElementRef, Input } from '@angular/core';
export class MdbBtnDirective {
    /**
     * @param {?} el
     * @param {?} renderer
     */
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.color = '';
        this.rounded = false;
        this.gradient = '';
        this.outline = false;
        this.flat = false;
        this.size = '';
        this.block = false;
        this.floating = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        const colorClass = 'btn-' + this.color;
        /** @type {?} */
        const gradientClass = this.gradient + '-gradient';
        /** @type {?} */
        const outlineClass = 'btn-outline-' + this.color;
        /** @type {?} */
        const flatClass = 'btn-flat';
        /** @type {?} */
        const roundedClass = 'btn-rounded';
        /** @type {?} */
        const sizeClass = 'btn-' + this.size;
        /** @type {?} */
        const blockClass = 'btn-block';
        /** @type {?} */
        const floatingClass = 'btn-floating';
        this.renderer.addClass(this.el.nativeElement, 'btn');
        if (this.color !== '') {
            this.renderer.addClass(this.el.nativeElement, colorClass);
        }
        if (this.rounded) {
            this.renderer.addClass(this.el.nativeElement, roundedClass);
        }
        if (this.gradient) {
            if (this.color !== '') {
                this.renderer.removeClass(this.el.nativeElement, colorClass);
            }
            this.renderer.addClass(this.el.nativeElement, gradientClass);
        }
        if (this.outline) {
            this.renderer.removeClass(this.el.nativeElement, colorClass);
            this.renderer.addClass(this.el.nativeElement, outlineClass);
        }
        if (this.flat) {
            if (this.color) {
                this.renderer.removeClass(this.el.nativeElement, colorClass);
            }
            if (this.gradient) {
                this.renderer.removeClass(this.el.nativeElement, gradientClass);
            }
            if (this.outline) {
                this.renderer.removeClass(this.el.nativeElement, outlineClass);
            }
            if (this.rounded) {
                this.renderer.removeClass(this.el.nativeElement, roundedClass);
            }
            this.renderer.addClass(this.el.nativeElement, flatClass);
        }
        if (this.size) {
            this.renderer.addClass(this.el.nativeElement, sizeClass);
        }
        if (this.block) {
            this.renderer.addClass(this.el.nativeElement, blockClass);
        }
        if (this.floating) {
            this.renderer.removeClass(this.el.nativeElement, 'btn');
            this.renderer.addClass(this.el.nativeElement, floatingClass);
        }
    }
}
MdbBtnDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mdbBtn]'
            },] }
];
/** @nocollapse */
MdbBtnDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
MdbBtnDirective.propDecorators = {
    color: [{ type: Input }],
    rounded: [{ type: Input }],
    gradient: [{ type: Input }],
    outline: [{ type: Input }],
    flat: [{ type: Input }],
    size: [{ type: Input }],
    block: [{ type: Input }],
    floating: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    MdbBtnDirective.prototype.color;
    /** @type {?} */
    MdbBtnDirective.prototype.rounded;
    /** @type {?} */
    MdbBtnDirective.prototype.gradient;
    /** @type {?} */
    MdbBtnDirective.prototype.outline;
    /** @type {?} */
    MdbBtnDirective.prototype.flat;
    /** @type {?} */
    MdbBtnDirective.prototype.size;
    /** @type {?} */
    MdbBtnDirective.prototype.block;
    /** @type {?} */
    MdbBtnDirective.prototype.floating;
    /**
     * @type {?}
     * @private
     */
    MdbBtnDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    MdbBtnDirective.prototype.renderer;
}
